sql/config.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package sql
+
+import (
+	"github.com/cockroachdb/cockroach/config"
+	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/encoding"
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+func init() {
+	// TODO(marc): we use a hook to avoid a depency on the sql package.
+	// We should probably move keys/protos elsewhere.
+	config.ZoneConfigHook = GetZoneConfig
+}
+
+// Given an object ID, build the key to its entry in the zones table.
+func makeZoneKey(id uint32) proto.Key {
+	k := keys.MakeTablePrefix(uint32(keys.ZonesTableID))
+	k = encoding.EncodeUvarint(k, uint64(ZonesTable.PrimaryIndex.ID))
+	k = encoding.EncodeUvarint(k, uint64(id))
+	k = encoding.EncodeUvarint(k, uint64(ZonesTable.Columns[1].ID))
+	return k
+}
+
+// GetZoneConfig returns the zone config for the object with 'id'.
+func GetZoneConfig(cfg *config.SystemConfig, id uint32) (*config.ZoneConfig, error) {
+	// Look in the zones table.
+	if val, ok := cfg.Get(makeZoneKey(id)); ok {
+		zone := &config.ZoneConfig{}
+		if err := gogoproto.Unmarshal(val, zone); err != nil {
+			return nil, err
+		}
+		// We're done.
+		return zone, nil
+	}
+
+	// No zone config for this ID. We need to figure out if it's a database
+	// or table. Lookup its descriptor.
+	rawDesc, ok := cfg.Get(MakeDescMetadataKey(ID(id)))
+	if !ok {
+		// No descriptor. This table/db could have been deleted,
+		// just return the default config.
+		return config.DefaultZoneConfig, nil
+	}
+
+	// Determine whether this is a database or table.
+	// TODO(marc): we need a better way of doing this. Options include:
+	// - add a type field on the descriptor table
+	// - separate descriptor tables for databases and tables
+	// - prebuild list of databases and tables in the system config
```

Comment:
You could also encode the type in the low-order bits of the ID. Reserving the low 2-4 bits for descriptor type seems reasonable and makes the ID self-describing. Definitely fix this (whatever the solution) in a separate change, though.